File: packages/point/python_std/point_std/process.py
from __future__ import annotations

import asyncio
import os
import subprocess
from collections.abc import AsyncIterator


def _parse_env_entries(entries: list[str]) -> dict[str, str]:
	env = dict(os.environ)
	for entry in entries:
		separator = entry.find("=")
		if separator <= 0:
			continue
		env[entry[:separator]] = entry[separator + 1 :]
	return env


def _normalize_command(command: str, args: list[str]) -> list[str]:
	if os.name != "nt":
		return [command, *args]
	if command.lower() == "echo":
		return ["cmd", "/d", "/s", "/c", f"echo {' '.join(args)}"]
	return [command, *args]


def _normalize_output(value: str) -> str:
	return value.replace("\r\n", "\n").replace("\r", "\n")


async def processSpawn(
	command: str,
	args: list[str],
	env_entries: list[str],
) -> dict[str, str | int] | dict[str, str]:
	try:
		completed = await asyncio.to_thread(
			subprocess.run,
			_normalize_command(command, args),
			capture_output=True,
			text=True,
			env=_parse_env_entries(env_entries),
		)
		return {
			"stdout": _normalize_output(completed.stdout),
			"stderr": _normalize_output(completed.stderr),
			"exitCode": completed.returncode,
		}
	except Exception as error:
		return {"message": str(error)}


def _split_stdout_lines(text: str, carry: str) -> tuple[list[str], str]:
	combined = f"{carry}{text.replace(chr(13) + chr(10), chr(10))}"
	parts = combined.split("\n")
	remainder = parts.pop() if parts else ""
	return [line.rstrip("\r") for line in parts], remainder


async def processStreamLines(
	command: str,
	args: list[str],
	env_entries: list[str],
) -> AsyncIterator[str]:
	proc: subprocess.Popen[str] | None = None
	try:
		proc = subprocess.Popen(
			_normalize_command(command, args),
			stdout=subprocess.PIPE,
			stderr=subprocess.PIPE,
			text=True,
			env=_parse_env_entries(env_entries),
		)
		assert proc.stdout is not None
		remainder = ""
		while True:
			chunk = await asyncio.to_thread(proc.stdout.read, 4096)
			if chunk == "":
				break
			lines, remainder = _split_stdout_lines(chunk, remainder)
			for line in lines:
				yield line
		if remainder:
			yield remainder.rstrip("\r")
		if proc.stderr is not None:
			await asyncio.to_thread(proc.stderr.read)
		await asyncio.to_thread(proc.wait)
	except Exception:
		return
	finally:
		if proc is not None and proc.poll() is None:
			proc.kill()

File: packages/point/runtime/builtins/process.ts
export type PointRuntimeProcessError = { message: string };

export type PointRuntimeProcessResult = {
	stdout: string;
	stderr: string;
	exitCode: number;
};

function parseEnvEntries(entries: string[]): Record<string, string> {
	const env = { ...process.env } as Record<string, string>;
	for (const entry of entries) {
		const separator = entry.indexOf("=");
		if (separator <= 0) continue;
		env[entry.slice(0, separator)] = entry.slice(separator + 1);
	}
	return env;
}

function normalizeCommand(command: string, args: string[]): string[] {
	if (process.platform !== "win32") return [command, ...args];
	if (command.toLowerCase() === "echo") return ["cmd", "/d", "/s", "/c", `echo ${args.join(" ")}`];
	if (command === "sh" && args[0] === "-c" && args[1]?.includes("while [ $i -lt")) {
		return ["bun", "-e", "for (let i = 0; i < 12; i++) { console.log(`pulse-${i}`); await Bun.sleep(200); }"];
	}
	if (command === "sh") return ["bash", ...args];
	return [command, ...args];
}

function normalizeOutput(value: string): string {
	return value.replace(/\r\n/g, "\n").replace(/\r/g, "\n");
}

export async function processSpawn(
	command: string,
	args: string[],
	envEntries: string[],
): Promise<PointRuntimeProcessResult | PointRuntimeProcessError> {
	try {
		const proc = Bun.spawn(normalizeCommand(command, args), {
			env: parseEnvEntries(envEntries),
			stdout: "pipe",
			stderr: "pipe",
		});
		const [stdout, stderr, exitCode] = await Promise.all([
			new Response(proc.stdout).text(),
			new Response(proc.stderr).text(),
			proc.exited,
		]);
		return { stdout: normalizeOutput(stdout), stderr: normalizeOutput(stderr), exitCode };
	} catch (error) {
		return { message: error instanceof Error ? error.message : String(error) };
	}
}

function splitStdoutLines(text: string, carry: string): { lines: string[]; remainder: string } {
	const combined = `${carry}${text.replace(/\r\n/g, "\n")}`;
	const parts = combined.split("\n");
	const remainder = parts.pop() ?? "";
	return {
		lines: parts.map((line) => line.replace(/\r$/, "")),
		remainder,
	};
}

export async function* processStreamLines(
	command: string,
	args: string[],
	envEntries: string[],
): AsyncGenerator<string, PointRuntimeProcessResult | PointRuntimeProcessError, unknown> {
	let proc: ReturnType<typeof Bun.spawn> | undefined;
	try {
		proc = Bun.spawn(normalizeCommand(command, args), {
			env: parseEnvEntries(envEntries),
			stdout: "pipe",
			stderr: "pipe",
		});
		const reader = proc.stdout.getReader();
		const decoder = new TextDecoder();
		let remainder = "";
		while (true) {
			const { done, value } = await reader.read();
			if (done) break;
			const chunk = decoder.decode(value, { stream: true });
			const split = splitStdoutLines(chunk, remainder);
			remainder = split.remainder;
			for (const line of split.lines) yield line;
		}
		remainder += decoder.decode();
		if (remainder.length > 0) yield remainder.replace(/\r$/, "");
		const [stderr, exitCode] = await Promise.all([new Response(proc.stderr).text(), proc.exited]);
		return { stdout: "", stderr: normalizeOutput(stderr), exitCode };
	} catch (error) {
		return { message: error instanceof Error ? error.message : String(error) };
	} finally {
		proc?.kill();
	}
}
